src/posts/2020-07-17-vim-elixir-ide/index.md

```diff
@@ -0,0 +1,248 @@
+It integrates with linters for pretty much every programming language. Here
+we're configuring it to use the `mix_format` fixer, which runs `mix format` on
+the current buffer whenever it is saved. Here it is happening in real time:
+
+![Vim Formatter](./formatter.gif)
+
+## Language Server
+
+Language Servers are all the rage nowadays.  They provide a common API for
+editors to understand each language's particularities, without having to resort
+to ad-hoc plugins that are never without their flaws.
+
+Language servers power tools such as the usual autocomplete list, "jump to
+defintion" and "find ocurrences" that Java and other OO-programmers are used to,
+but in a more systematic and language-agnostic way.
+
+The language server implementation for Elixir is [`elixir_ls`][elixir_ls] and
+for Vim can be integrated with [`coc.nvim`], the go-to language server client:
+
+```vimscript
+Plug 'neoclide/coc.nvim', { 'branch': 'release' }
+Plug 'amiralies/coc-elixir', {'do': 'yarn install && yarn prepack'}
+```
+
+This gives you several new superpowers:
+
+### 1. Inline dyalizer warnings
+
+[Dialyzer][dialixir], a static analysis tool for Erlang/Elixir is, by itself,
+a very useful but cumbersome tool to use.  It often takes too long to run, and
+ends up relegated to being that tool you only run every couple of weeks, only to
+find out you've been piling up on those warnings from the type system.
+
+Having your editor run it for you in the background, and allowing you to have
+near-instant feedback, makes the entire process much better.
+
+
+![Dialyer warnings](./dialyzer-warnings.png) <div style="margin-top: 5em"></div>
+
+### 2. Semantic code navigation
+
+```vimscript
+nmap <silent> gd <Plug>(coc-definition)
+nmap <silent> gr <Plug>(coc-references)
+```
+
+Shorcuts such as this (and many more which you can check on the [help
+page][coc-help]) provides quick contextual navigation within Elixir functions.
+
+![Showing references](./references.gif) <div style="margin-top: 5em"></div>
+
+### 3. View documentation on the fly
+
+```vimscript
+nnoremap <silent> K :call <SID>show_documentation()<CR>
+```
+
+![Documentation](./documentation.gif)
+<div style="margin-top: 5em"></div>
+
+### 4. View an outline of current file
+
+```vimscript
+nnoremap <silent> <leader>co  :<C-u>CocList outline<CR>
+```
+
+![Outline](./outline.gif)
+<div style="margin-top: 5em"></div>
+
+## Project-wide Find/Replace
+
+I've been searching for a good find & replace solution for Vim for a long time.
+(dare I say, I've been finding & replacing them all this time?)
+
+Recently, I found my perfect one so far, in the form of a very simple and hacky
+trick: [quickfix-reflector][quickfix-reflector].
+
+The quickfix window is a special non-editable buffer, usually used display
+things like log messages or compiler errors.  It can also be used to show search
+results, such as when calling `:Rg some_function` to find all ocurrences of
+`some_function` in a codebase.
```

Comment:
Feel free to ignore this but I'd probably mention vim-abolish due to adding `:%S` in there

Reply:
Doesn't feel right at all to include something I don't use myself

Reply:
That's fair so feel free to ignore it